drivers/net/ethernet/intel/idpf/eth/idpf_eth.c

```diff
@@ -0,0 +1,380 @@
+				   &adapter->post_init_task,
+				   msecs_to_jiffies(1 * (pdev->devfn & 0x07)));
+		return;
+	}
+
+	/* Ready to finish the post init now */
+	idpf_eth_device_post_init(adapter);
+}
+
+/**
+ * idpf_eth_device_pre_init - Ethernet device pre-initialization routine
+ * @adapter: Ethernet private data struct
+ */
+static int idpf_eth_device_pre_init(struct idpf_eth_adapter *adapter)
+{
+	struct idpf_vport_config *vport_config;
+	struct idpf_eth_idc_dev_info *dev_info;
+	struct idpf_eth_shared *eth_shared;
+	struct idpf_netdev_priv *np;
+	struct idpf_vport *vport;
+	struct pci_dev *pdev;
+	int err;
+
+	dev_info = adapter->dev_info;
+	eth_shared = idpf_eth_adapter_shared(adapter);
+	pdev = eth_shared->pdev;
+	adapter->req_tx_splitq = true;
+	adapter->req_rx_splitq = true;
+	err = idpf_send_create_vport_msg(adapter, &dev_info->caps.q_info);
+	if (err)
+		goto unwind_vports;
+
+	vport = idpf_vport_alloc(adapter, &dev_info->caps.q_info);
+	if (!vport) {
+		err = -EFAULT;
+		dev_err(&pdev->dev, "failed to allocate vport: %d\n",
+			err);
+		goto unwind_vports;
+	}
+
+	vport_config = adapter->vport_config;
+	init_waitqueue_head(&vport->sw_marker_wq);
+	spin_lock_init(&vport_config->mac_filter_list_lock);
+	INIT_LIST_HEAD(&vport_config->user_config.mac_filter_list);
+
+	err = idpf_check_supported_desc_ids(vport);
+	if (err) {
+		dev_err(&pdev->dev, "failed to get required descriptor ids\n");
+		goto cfg_netdev_err;
+	}
+
+	if (idpf_cfg_netdev(vport))
+		goto cfg_netdev_err;
+
+	err = idpf_send_get_rx_ptype_msg(vport);
+	if (err)
+		goto handle_err;
+
+	/* Once state is put into DOWN, driver is ready for dev_open */
+	np = netdev_priv(vport->netdev);
+	np->state = __IDPF_VPORT_DOWN;
+	if (test_and_clear_bit(IDPF_VPORT_UP_REQUESTED, vport_config->flags))
+		idpf_vport_open(vport, true);
+
+	return 0;
+
+handle_err:
+	idpf_decfg_netdev(vport);
+cfg_netdev_err:
+	idpf_vport_rel(vport);
+	adapter->vport = NULL;
+unwind_vports:
+	if (adapter->vport)
+		idpf_vport_dealloc(adapter->vport);
+	return err;
+}
+
+/**
+ * idpf_eth_probe - Probe ethernet device
+ * @adev: Structure related to auxiliary ethernet device
+ * @id: auxiliary device id
+ */
+static int idpf_eth_probe(struct auxiliary_device *adev,
+			  const struct auxiliary_device_id *id)
+{
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	struct idpf_eth_adapter *adapter;
+	struct pci_dev *pdev;
+	int err;
+
+	eth_dev = container_of(adev, struct idpf_eth_idc_auxiliary_dev, adev);
+	adapter = dev_get_drvdata(&adev->dev);
+	pdev = eth_dev->eth_info.eth_shared->pdev;
+	if (!adapter) {
+		adapter = idpf_eth_adapter_alloc(pdev);
+		if (!adapter)
+			return -ENOMEM;
+
+		/* Initialize dev_info */
+		adapter->dev_info = &eth_dev->eth_info;
+		dev_set_drvdata(&adev->dev, adapter);
+
+	}
+
+	/* Save Eth device info */
+	adapter->dev_info = &eth_dev->eth_info;
+	adapter->dev_info->start_probe_post_init = false;
+
+	/* Device's pre-initialization */
+	err = idpf_eth_device_pre_init(adapter);
```

Comment:
I don't see the probe as divided into pre-init, probe and then post-init.
I am not seeing the value of this kind of division. Its just one probe . The pre-init, if any is in the main driver when it allocates the aux device, core dev info, initializes them and does auxiliary_device_init and 'add'.

Reply:
It is done on purpose to keep the vport registration serial. As you know all vport registration in parallel doesn't work.  I think we need some kind of logic to make it happen. And I came up with this.